### PR TITLE
[wip] Graph backend

### PR DESCRIPTION
**Pull request by viperscape**
Work in progress, wanted to share where it's at. I recently added an edge guard which needs more work, bit confusing right now-- but I think it will work for the blend graph needs. The blend graph's float parameter can come from the edge factor, see the test for some examples.

### CODE REVIEW

**Comment by viperscape** [comment]
So some things to think about for this backend would be to:
- [ ] Finish up a trait for the Graph impl as GraphBase trait so the backend can be agnostic. 
- [ ] Moving away from using NodeBase trait as inherited and instead use a trait object, like SourceData in your example. 
- [ ] Clean up the edge guard syntax, and extend it. 
- [ ] Finish the Edge-factor trait fn, for specialized 'weighted' edges that could be different than float
- [ ] Possibly create a new trait for node state transformations (like a mixer between output nodes, eg. your AnimClips and float params)
- [ ] Create weighted, directed, and maybe tracking traits for graph types. It might be best to make graph tracking connections default.

**Comment by viperscape** [comment]
This is now working like the original before it. Probably time for a review and possible merge. When you have a chance @stjahns :)

File: src/graph.rs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-use conrod::{Ui, UiId, Button, Label, Positionable, Labelable, Sizeable};
```
Comment (stjahns) [inline src/graph.rs:1]: This file can be deleted, right?

**Comment by stjahns** [comment]
Looking good! I think this looks good to merge. I think I understand where you're going with the EdgeGuard trait! We'll eventually have to figure out how to draw 'In/Out' nodes as part of a 'Root' node, maybe another trait that defines `is_root() -> bool` ? Then we only draw the full node widget for root nodes, and draw all connected non-root nodes within that widget?

**Comment by stjahns** [comment]
Sorry I'm not able to spend a huge deal of time on this anymore, hopefully I'll get more of a chance soonish

**Comment by viperscape** [comment]
no worry, I'm in no major rush any help is appreciated :)
graph can be deleted, or optionally pull out graph impl from backend and place in graph.rs

is_root is a good idea. More work on edge guards is likely, and figuring out how to render that nicely :)

**Pull request merged**